### PR TITLE
add: fabric subscription protocol

### PR DESCRIPTION
**Pull request by vinliao**
# Thank you for contributing to Spellbook!

Thank you for taking the time to submit code in Spellbook. A few things to consider:

- If you are a first-time contributor, please sign the CLA by copy & pasting **_exactly_** what the bot mentions in PR comment
- Refer to [docs](#spellbook-contribution-docs) section below to answer questions
- Dune team will review submitted PRs as soon as possible

## Spellbook sub-projects

To prepare for Spellbook to scale in size, Dune team has started to pilot sub-projects:

- For a quick overview, see [the GH discussion on the topic](https://github.com/duneanalytics/spellbook/discussions/5238)
  - Please ask questions and continue the conversation there
- The first spell lineage to move: `tokens.erc20` and all upstream spells which help build it
  - **_Note_**: Please navigate to [tokens](/tokens) root directory to find tokens objects

## Spellbook contribution docs

The [docs](docs) directory has been implemented to answer as many questions as possible. Please take the time to reference each `.md` file within this directory to understand how to efficiently contribute & why the repo is designed as it is 🪄

Example questions to be answered:

- What does each property in the [model config block](/docs/models/model_config_block.md) mean?
- What is the [CI test](/docs/ci_test/ci_test_overview.md) attached to PRs and how can I best utilize it?
- What are the Spellbook [best practices](/docs/general/best_practices.md)?

Please navigate through the [docs](/docs) directory to find as much info as you can.

**Note**: happy to take PRs to improve the docs, let us know 🤝

### CODE REVIEW

**Comment by vinliao** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by vinliao** [comment]
Hello, I'm trying to add queries related to Fabric (https://docs.withfabric.xyz/stp/overview).

I do have several questions: 
- where does the seed CSV come from? (seed/_sector/referral/{name}.csv)
- what is currency_contract in macros? ,{{ var("ETH_ERC20_ADDRESS") }} as currency_contract

Is there anything else lacking to make this PR work?

**Comment by vinliao** [comment]
Anything I can do on my end to help move this PR forward? @0xRobin

**Comment by 0xRobin** [comment]
@vinliao, looking at the current CI error..
You'll need to add the sources you're using here:
https://github.com/duneanalytics/spellbook/blob/58938f31f355bd032c453c8456ef83b7bb816ce8/sources/_sector/referral/rewards/platforms/_sources.yml

**Comment by 0xRobin** [comment]
> * what if the currency contract info (macros) is not available in the table (in this case: SubscriptionTokenV1_evt_ReferralPayout)

If you can't source the currency contract from this event, you might be able to source it from the subscriptiontoken creation event or function call and join it in? I don't recall the exact function/event but I think we were talking about this when you worked on the initial dashboard.

**Comment by vinliao** [comment]
image

Okay, I'm seeing this. Could you give me pointers on how to fix this? Here's the macro:

```
{% macro fabric_referral_rewards(
    blockchain
    ,SubscriptionTokenV1_evt_ReferralPayout
    ,SubscriptionTokenV1Factory_call_deploySubscription
    )
%}

with currency_contract_info as (
    select
        erc20TokenAddr,
        contract_address
    from {{SubscriptionTokenV1Factory_call_deploySubscription}}
)

select
    '{{blockchain}}' as blockchain
    ,'fabric' as project
    ,'v1' as version
    ,evt_block_number as block_number
    ,evt_block_time as block_time
    ,cast(date_trunc('day',evt_block_time) as date) as block_date
    ,cast(date_trunc('month',evt_block_time) as date) as block_month
    ,evt_tx_hash as tx_hash
    ,'NFT' as category
    ,referrer as referrer_address
    ,tx."from" as referee_address
    ,c.erc20TokenAddr as currency_contract
    ,"rewardAmount" as reward_amount_raw
    ,e."contract_address" as project_contract_address
    ,evt_index as sub_tx_id
    ,tx."from" as tx_from
    ,tx.to as tx_to
from {{SubscriptionTokenV1_evt_ReferralPayout}} e
inner join {{source(blockchain, 'transactions')}} tx
    on e.evt_block_number = tx.block_number
    and e.evt_tx_hash = tx.hash
left join currency_contract_info c
    on e.contract_address = c.contract_address
    {% if is_incremental() %}
    and {{incremental_predicate('tx.block_time')}}
    {% endif %}
{% if is_incremental() %}
where {{incremental_predicate('evt_block_time')}}
{% endif %}
{% endmacro %}
```

erc20TokenAddr is `0x0000...` if it's Eth, else it's the currency's contract address. erc20TokenAddr is NOT NULL in the original table (fabric_base.SubscriptionTokenV1Factory_call_deploySubscription). Could you give me pointers on what I'm doing wrong?

File: macros/models/_sector/referral/platforms/fabric_referral_rewards.sql [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    blockchain
+    ,SubscriptionTokenV1_evt_ReferralPayout
+    ,SubscriptionTokenV1Factory_call_deploySubscription
+    )
+%}
+
+with currency_contract_info as (
+    select
+        erc20TokenAddr,
+        contract_address
```
Comment (0xRobin) [inline macros/models/_sector/referral/platforms/fabric_referral_rewards.sql:11]: You need to take the `output_0` here.
The contract_address is the factory itself

File: macros/models/_sector/referral/platforms/fabric_referral_rewards.sql [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    ,tx.to as tx_to
+from {{SubscriptionTokenV1_evt_ReferralPayout}} e
+inner join {{source(blockchain, 'transactions')}} tx
+    on e.evt_block_number = tx.block_number
+    and e.evt_tx_hash = tx.hash
+left join currency_contract_info c
+    on e.contract_address = c.contract_address
+    {% if is_incremental() %}
+    and {{incremental_predicate('tx.block_time')}}
+    {% endif %}
```
Comment (0xRobin) [inline macros/models/_sector/referral/platforms/fabric_referral_rewards.sql:41]: the incremental block should be moved up some lines so it is in the inner join condition with transactions.

**Review by 0xRobin — CHANGES_REQUESTED** [review]
this should fix it. :pray:

File: macros/models/_sector/referral/platforms/fabric_referral_rewards.sql (hunk 1 above)
Comment (vinliao) [reply macros/models/_sector/referral/platforms/fabric_referral_rewards.sql]: Thanks, these two comments makes the CI green.

**Comment by vinliao** [comment]
image


Green CI and it works on dune.com, thank you very much 🙏

Some questions:
- will it auto-add the data to the referral-market dashboard when it's merged?
- just to triple check: the `SubscriptionTokenV1_evt_ReferralPayout` table lists out the referral reward on Fabric STP (if I refer Alice, I will get a percentage if Alice purchases a subscription). This is what the referral-market Dune dashboard tracks right?

Anything else missing before this PR can be merged?

**Comment by vinliao** [comment]
Let me know if there's anything lacking before this PR can be merged @0xRobin :)

**Review by jeff-dude — APPROVED** [review]

**Pull request merged**